fms/models/roberta.py

```diff
@@ -0,0 +1,335 @@
+            mask: torch.BoolTensor = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+
+        x_emb = self.embedding(x)
+
+        # if pad_id exists
+        #   is_pad will be a BoolTensor
+        #   otherwise pad_id will not be taken into account
+        if self.config.pad_id is None:
+            is_pad = torch.zeros_like(x, dtype=bool, device=x.device)
+        else:
+            is_pad = x == self.config.pad_id
+
+        if position_ids is None:
+            position_ids = ((~is_pad).cumsum(1) - 1).clamp(min=0)
+
+        # look up position embeddings
+        position_out = self.position_embedding(position_ids)
+
+        # zero out the associated position embeddings
+        if self.config.pad_id is not None:
+            position_out = position_out.mul(~is_pad.unsqueeze(-1))
+
+        # perform absolute position embedding
+        x = x_emb + position_out
+
+        # layer norm
+        x = self.enc_norm(x)
+
+        # add dropout
+        if self.config.p_dropout:
+            x = self.dropout(x)
+
+        # layers
+        for layer in self.layers:
+            x = layer(x, mask=mask, attn_algorithm=attn_algorithm)
+
+        return x
+
+
+class RoBERTa(nn.Module):
+    def __init__(self, config: Optional[RoBERTaConfig] = None, **kwargs):
+
+        super(RoBERTa, self).__init__()
+        if config is not None:
+            self.config = config
+        else:
+            self.config = RoBERTaConfig()
+        self.config = self.config.updated(**kwargs)
+
+        self.base_model = RoBERTaHeadless(self.config)
+
+        self.class_head = ClassHead(
+            self.config.emb_dim,
+            str_to_activation(self.config.activation_fn),
+            self.config.norm_eps,
+        )
+
+        self.head = nn.Linear(self.config.emb_dim, self.config.src_vocab_size)
+
+        # this model ties weights, so we tie here
+        self.head.weight = self.base_model.embedding.weight
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        attn_algorithm: Optional[str] = None,
+    ):
+        # run through the encoder layers
+        x = self.base_model(
+            x, mask=mask, position_ids=position_ids, attn_algorithm=attn_algorithm
+        )
+
+        # run through the class head (using the first in each sequence in the batch as the cls_token)
+        x = self.class_head(x)
+
+        # project to vocab space
+        x = self.head(x)
+        return x
+
+    @classmethod
+    def from_config(cls, config: RoBERTaConfig) -> "RoBERTa":
+        return cls(config)
+
+    def get_config(self) -> RoBERTaConfig:
+        return self.config
+
+    def reset_params(self):
+        # Modules are self-initializing, we're just going to down-scale the final prediction head to be
+        # mixed-fan (inputs and gradients scale to the same inverse factors) if it isn't tied
+        self.head.weight.data.normal_(
```

Comment:
Given that weights are explicitly tied above, wouldn't this also overwrite the embedding weights?

Reply:
yes I believe this would. Maybe we should be doing:

```python
for layer in ["position_embedding", "embedding"]:
     nn.init.normal_(getattr(self.base_model, layer).weight, mean=0.0, std=self.config.emb_dim**-0.5)
self.head.bias.data.zero_()
```

Reply:
updated this in latest commit, the headless Roberta is now doing the nn.init.nomal_, and the roberta with head is doing the head bias zero